### DISCUSSION TITLE
How can i use a mode with another comparision?

### DISCUSSION THREAD

**Discussion by char-ptr** [Brendonovich/prisma-client-rust#93] [Q&A] [answered]
Im trying to lookup a string which is supposed to be case insensitive, in standard prisma you would be able to do something like the following:
```ts
prisma.user.findFirst({
where:{
name:{
mode:"insensitive",
equals:"xyz.."
}
}
})
```
i have now tried to replicate the following in rust, but i don't believe it's possible to chain it hence why it's not working.
```rs
prisma.user().find_first(vec![
        prisma::user::name::mode(prisma::_prisma::QueryMode::Insensitive),
        prisma::user::name::equals(name)
    ]).with(....).exec().await.unwrap()
```

**Comment by Brendonovich** [comment, marked as answer]
That's interesting, I haven't tested case sensitivity but the code you provided should work. I don't understand what you mean about chaining though?

**Reply by char-ptr** [reply]
Oh that's odd because the code i provided didn't function as intended unfortunately :(. and by chaining i mean like well using multiple options per column (sorry if explained poorly).

**Reply by Brendonovich** [reply]
Ah ok, I understand. If the code doesn't work then yeah I need to test it and fix it.

**Reply by Brendonovich** [reply]
I think it may have something to do with how I was using `transform_equals`, which may have interfered with the `mode` parameter. Can you try again using `branch = "0.6.0"` instead of `tag = "0.5.2"`?

**Reply by char-ptr** [reply]
i switched it over to that branch and i seem to still be having issues.

**Reply by Brendonovich** [reply]
Would you mind sharing the errors you are getting? I can't really help without seeing what's going wrong.

**Reply by char-ptr** [reply]
I'm not getting any errors, im just not getting a result found. i'll share the rust code again here.
```rs
//failing result
use prisma::PrismaClient;
use prisma_client_rust::NewClientError;

pub mod prisma;
#[tokio::main]
async fn main() {
    dotenv::dotenv().ok();
    let client: Result<PrismaClient, NewClientError> = prisma::new_client().await;
    if let Ok(client) = &client {
        let d = client.user().find_first(vec![
            prisma::user::name::mode(prisma::_prisma::QueryMode::Insensitive),
            prisma::user::name::equals("luna".to_string())
    
        ]).exec().await.unwrap();
        println!("{:?}", d);
    } else if let Err(e) = &client {
        println!("{:?}", e);
    }
}
```
```rs
// gets result
use prisma::PrismaClient;
use prisma_client_rust::NewClientError;

pub mod prisma;
#[tokio::main]
async fn main() {
    dotenv::dotenv().ok();
    let client: Result<PrismaClient, NewClientError> = prisma::new_client().await;
    if let Ok(client) = &client {
        let d = client.user().find_first(vec![
            prisma::user::name::mode(prisma::_prisma::QueryMode::Insensitive),
            prisma::user::name::equals("Luna".to_string())
    
        ]).exec().await.unwrap();
        println!("{:?}", d);
    } else if let Err(e) = &client {
        println!("{:?}", e);
    }
}
```
The difference is from changing `luna` to be uppercase, but with mode set to insensitive it should work regardless of the case.

**Reply by char-ptr** [reply]
Any update on this?

**Reply by Brendonovich** [reply]
Been on my radar but I haven't looked into it just yet. Will make a tracking issue to make sure I get it checked out for 0.6.0.

**Reply by Brendonovich** [reply]
@pozm I think I've fixed it, turns out I wasn't merging the filters properly. Try 0.6.0, but read what I've added to the setup instructions first. There's a bit more stuff to pay attention to during setup now 🙃

**Reply by char-ptr** [reply]
Unfortunate still seems not to be working.
output:
```
all users: Ok([Data { id: "cl58rg6qh0015zg9mr88dv1ze", name: "Luna", api_key: "cl58rg6qh0014zg9m8swmybkg", activity: None }])
search for 'Luna' : Some(Data { id: "cl58rg6qh0015zg9mr88dv1ze", name: "Luna", api_key: "cl58rg6qh0014zg9m8swmybkg", activity: None })
search for 'luna' : None
```
src
```rust
use crate::prisma::_prisma::QueryMode;
//....
            println!("all users: {:?}", client.user().find_many(vec![]).exec().await);
            let d = client.user().find_first(vec![
                prisma::user::name::mode(QueryMode::Insensitive),
                prisma::user::name::equals("Luna".to_string())
        
            ]).exec().await.unwrap();
            println!("search for 'Luna' : {:?}", d);
            let d = client.user().find_first(vec![
                prisma::user::name::mode(QueryMode::Insensitive),
                prisma::user::name::equals("luna".to_string())
        
            ]).exec().await.unwrap();
            println!("search for 'luna' : {:?}", d);

```

I also tried using stable (cause previously i was using nightly) but hasn't made a difference,

**Reply by Brendonovich** [reply]
Just to be _totally_ sure, try with `rev = "ee0d7d3efebb450ce3275cd06a123d45b5c6eed5"` instead of `branch = "0.6.0"`. If PCR didn't update properly then that should fix it, but if not then I'll be very confused and would appreciate a repro repo.

**Reply by char-ptr** [reply]
Unfortunately it still doesn't work. i also decided to fork and add an additional test for insensitive finds, it turns out that it might be something at the engine level? as the operation thingy seems to have merged everything fine, here was the output
```rust
[src\queries\find_first.rs:163] Operation::Read(selection.build()) = Read(
    Selection {
        name: "findFirstPost",
        alias: Some(
            "result",
        ),
        arguments: [
            (
                "where",
                Object(
                    {
                        "title": Object(
                            {
                                "equals": String(
                                    "test post 1",
                                ),
                                "mode": Enum(
                                    "insensitive",
                                ),
                            },
                        ),
                    },
                ),
            ),
        ],
        nested_selections: [
            Selection {
                name: "id",
                alias: None,
                arguments: [],
                nested_selections: [],
            },
            Selection {
                name: "created_at",
                alias: None,
                arguments: [],
                nested_selections: [],
            },
            Selection {
                name: "updated_at",
                alias: None,
                arguments: [],
                nested_selections: [],
            },
            Selection {
                name: "title",
                alias: None,
                arguments: [],
                nested_selections: [],
            },
            Selection {
                name: "published",
                alias: None,
                arguments: [],
                nested_selections: [],
            },
            Selection {
                name: "views",
                alias: None,
                arguments: [],
                nested_selections: [],
            },
            Selection {
                name: "desc",
                alias: None,
                arguments: [],
                nested_selections: [],
            },
            Selection {
                name: "author_id",
                alias: None,
                arguments: [],
                nested_selections: [],
            },
        ],
    },
)
```
the reproduce repo can be found [here](https://github.com/pozm/prisma-insensitive-test)
and the fork can be found [here](https://github.com/pozm/prisma-client-rust) (6.0.0 branch)

**Reply by Brendonovich** [reply]
I'm suddenly not convinced this is a PCR problem 😅 
<img width="857" alt="image" src="https://user-images.githubusercontent.com/14191578/184884449-3a594b39-6f8d-4cb2-a24d-9001d72bf2c1.png">

I tested with a fresh Postgres image in docker, maybe it's a PG version problem?
<img width="396" alt="image" src="https://user-images.githubusercontent.com/14191578/184885464-69c288e2-0ecf-49f8-b652-f0501ce7a407.png">

Also I made a [couple of changes](https://github.com/Brendonovich/prisma-insensitive-test/commit/8a561a5ed3c9e58908957421bb0f9197f629f866) to your repo with some QoL improvements 😄

**Reply by char-ptr** [reply]
I was already using postgres 14.3 (installed via scoop, globally), although oddly enough it does now seem to work when using one from dockerhub. how very strange. i just checked again, and for some reason when i use the old database (from scoop) with prisma for js, it works. i wonder what's going on here?

i made another test repository for the ts code i used, which can be found [here](https://github.com/pozm/prisma-js-insensitive-case)

**Reply by Brendonovich** [reply]
No idea why it's not working on your end, but from what I've tested PCR does support insensitive querying now. Going to close the issue and mark as fixed, but if you have problems with other version of Postgres then please let me know.

**Reply by char-ptr** [reply]
Alright thank you so much for helping me through this
